sympy/solvers/solveset.py

```diff
@@ -231,6 +235,26 @@ def _invert_complex(f, g_ys, symbol):
                                for g_y in g_ys if g_y != 0])
             return _invert_complex(f.args[0], exp_invs, symbol)
 
+    if isinstance(f, TrigonometricFunction):
+        if isinstance(g_ys, FiniteSet):
+            def inv(trig):
+                if isinstance(f, (sin, csc)):
+                    F = asin if isinstance(f, sin) else acsc
+                    return (lambda a: n*pi + (-1)**n*F(a),)
+                if isinstance(f, (cos, sec)):
+                    F = acos if isinstance(f, cos) else asec
+                    return (
+                        lambda a: 2*n*pi + F(a),
+                        lambda a: 2*n*pi - F(a),)
+                if isinstance(f, (tan, cot)):
+                    return (lambda a: n*pi + f.inverse()(a),)
+
+            n = Dummy('n', integer=True)
+            invs = S.EmptySet
+            for L in inv(f):
+                invs += Union(*[imageset(Lambda(n, L(g)), S.Integers) for g in g_ys])
+            return _invert_real(f.args[0], invs, symbol)
+    #TODO similarly for HyperbolicFunction
```

Comment:
hyperbolics can be rewritten in terms of trig with the Osborne transformation and should be handled exactly the same way as the trig expressions. Otherwise you start to get diverging behavior in how the two respond to attempts at solution.